typescript/sdk/src/token/EvmWarpCrudModule.ts

```diff
@@ -0,0 +1,216 @@
+  }
+
+  /**
+   * Retrieves the token router configuration for the specified address.
+   *
+   * @param address - The address to derive the token router configuration from.
+   * @returns A promise that resolves to the token router configuration.
+   */
+  public async read(): Promise<DerivedTokenRouterConfig> {
+    return this.reader.deriveWarpRouteConfig(
+      this.args.addresses.deployedWarpRoute,
+    );
+  }
+
+  /**
+   * Updates the Warp Route contract with the provided configuration.
+   *
+   * @remark Currently only supports updating ISM or hook
+   *
+   * @param config - The configuration for the token router to be updated.
+   * @returns An array of Ethereum transactions that were executed to update the contract, or an error if the update failed.
+   */
+  public async update(
+    config: DerivedTokenRouterConfig,
+  ): Promise<EthersV5Transaction[]> {
+    return [
+      ...(await this.updateIsm(config)),
+      ...(await this.updateHook(config)),
+    ];
+  }
+
+  /**
+   * Updates an existing Warp route ISM with a given configuration.
+   *
+   * This method handles two cases:
+   * 1. If the `config.interchainSecurityModule` is an object,
+   *  - Checks if the current onchain ISM configuration matches the provided configuration.
+   *  - If not, it deploys a new ISM module, and updates the contract's ISM.
+   * 2. If the `config.interchainSecurityModule` is a string
+   *  - It attempts to derive the ISM from the provided string, and updates the contract's ISM.
+   *
+   * @param config - The token router configuration, including the ISM configuration.
+   * @returns An array of Ethereum transactions that need to be executed to update the ISM configuration.
+   */
+  async updateIsm(
+    config: DerivedTokenRouterConfig,
+  ): Promise<EthersV5Transaction[]> {
+    const transactions: EthersV5Transaction[] = [];
+
+    const contractToUpdate = await this.args.addresses[config.type].deployed();
+
+    if (typeof config.interchainSecurityModule === 'string') {
+      // Derive & set ISM
+      const ism = await this.reader.evmIsmReader.deriveIsmConfig(
+        config.interchainSecurityModule,
+      );
+      transactions.push({
+        transaction:
+          await contractToUpdate.populateTransaction.setInterchainSecurityModule(
+            ism.address,
+          ),
+        type: ProviderType.EthersV5,
+      });
+    } else if (typeof config.interchainSecurityModule === 'object') {
+      const onchainConfig = await this.read();
+      if (
+        config.interchainSecurityModule.type !==
+        onchainConfig.interchainSecurityModule!.type
+      ) {
+        // Deploy & set ISM
+        const ismModule = await this.deployIsm(config);
+        transactions.push({
+          transaction:
+            await contractToUpdate.populateTransaction.setInterchainSecurityModule(
+              ismModule,
+            ),
+          type: ProviderType.EthersV5,
+        });
+      }
+    }
```

Comment:
```suggestion
    const onchain = await this.read();
    const expectedIsm = config.interchainSecurityModule;
    const actualIsm = onchain.interchainSecurityModule;
    const matchesIsm = typeof expectedIsm === 'string' ? eqAddress(expectedIsm, actualIsm.address) : deepEquals(expectedIsm, actualIsm);
    if (!matchesIsm) {
      transactions.push(await contractToUpdate.populateTransaction.setInterchainSecurityModule(expectedIsm.address));
    }
```
something like this?
might want to merge with the hook update to avoid `this.read()` twice

Reply:
```typescript
async updateIsm(
    config: DerivedTokenRouterConfig,
  ): Promise<EthersV5Transaction[]> {
    const transactions: EthersV5Transaction[] = [];
    if (config.interchainSecurityModule) {
      const contractToUpdate = await this.args.addresses[
        config.type
      ].deployed();

      // Attempt to derive the config.interchainSecurityModule if its a string, or deploy it if its an object
      const expectedIsm =
        typeof config.interchainSecurityModule === 'string'
          ? await this.reader.evmIsmReader.deriveIsmConfig(
              config.interchainSecurityModule,
            )
          : await this.deployIsm(config);

      // read() will return undefined if no ISM set, so make it an IsmConfig-ish object
      const actualIsm = (await this.read()).interchainSecurityModule ?? {
        address: ethers.constants.AddressZero, 
      };

      if (!deepEquals(expectedIsm, actualIsm)) {
        transactions.push({
          transaction:
            await contractToUpdate.populateTransaction.setInterchainSecurityModule(
              expectedIsm.address,
            ),
          type: ProviderType.EthersV5,
        });
      }
    }

    return transactions;
  }
```

Reply:
Fixed.